### ISSUE TITLE
Mock-up Stable Diffusion UI

### ISSUE THREAD

**Issue by angelmankel** [NickLucche/stable-diffusion-react-ui#1]
Hi Nick, here is what I have so far for the main UI. Let me know if you see anything that should be changed or if you have some other ideas I did not include. Thanks!

Notable Initial features:
- Prompt creator panel with history (not in the screenshots yet)
- Image gallery which allows users to favorite images, tag, and filter (not in screenshots)
- Recent images panel which allows a user to quickly preview, favorite, and recall parameters
- Easy UI color theme options
- Save prompt button to tag and add to prompt creator favorites
- Settings window that will allow users to set default values, change increment values for sliders, etc.
- Collapsable recall panel

Featues not initially planned but possible if the backend allows it:
- Img2img features & image editing such as masking, filters, cropping, inpainting, outpainting, etc.
- Allow user to send an image from the gallery to img2img to use as an input image

![Mock-up UI](https://user-images.githubusercontent.com/23511169/193726620-d52e5d34-313d-4b49-853f-06554f8a629f.png)
![UI with Notes](https://user-images.githubusercontent.com/23511169/193726624-4c18d1af-baae-475e-850b-3c58dcbdcd49.png)

**Comment by NickLucche** [comment]
Hey, this surely looks soo much better than the Gradio UI!
I think there are many interesting ideas that I didn't know about such as promptmania style of building up text prompts.

I was wondering, should we add something backend side to support favorites or where you thinking about storing it in cache?
Face restoration but I think most importantly image upscaling (with SuperResolution networks) could definitely be added.
Same thing for img2img which is already on the todos list, I just need to set aside some time to work on it.
I believe things like masking, image in and outpainting would be so cool to have.

So yeah pretty much all the features you added would be nice to have and support from backend side.
Perhaps we could replace image restoration with image upscaling for the initial version?
Also, what do you mean here with variation seed and strength? Maybe I missed it if it is already featured somewhere.

Thanks for working on it!

**Comment by angelmankel** [comment]
> Hey, this surely looks soo much better than the Gradio UI! I think there...

Thank you! Yeah I recently came across that website and I think something like that would make prompt building much faster/easier.

> I was wondering, should we add something backend side to support favorites or where you thinking about storing it in cache?

So I've been thinking about that and i'm still a bit undecided. We could either A: store persistent data using the Web Storage API, B: include a SQL server in the docker container which the UI can connect to, or C: store a SQLite database file locally to avoid having to spin up a second docker container for SQL. I haven't spent much time utilizing the Web Storage API for potentially thousands of images so i'm not sure how it would hold up, but an SQL db should work just fine. Maybe you have a preference/opinion about this but right now i'm leaning towards B.

For the gallery to work correctly I think storing the images in a database rather than programmatically looking through a file directory would allow for much more efficient filtering/searching/etc. Each row could contain the image in base64 along with other columns for favorites, tags, parameters, or whatever we want.

> Same thing for img2img which is already on the todos list, I just need to set aside some time to work on it.

Didn't know Img2img was planned so my mistake. But that's great news!

> Perhaps we could replace image restoration with image upscaling for the initial version?

Sure, I can restructure a bit to include upscaling instead of restoration. But it might be worth just making a post processing tab which can include upscaling, restoration, etc. to clear up space in the editor. We could right click an image in the gallery/recents and 'send' to the post processing tab. It can also include a more condensed gallery on the side. It would work similarly to this one: https://github.com/AUTOMATIC1111/stable-diffusion-webui

> Also, what do you mean here with variation seed and strength?

Variation seed is a feature that is included in a few stable-diffusion repos. Basically, you have a second seed which you can blend into the initial seed. Strength (0-1 range) just tells SD how much to blend. I don't know anything about the programming logic for this and I'm not sure if this is something in SD by default but I will look around and see if I can find out how it can be implemented.

**Another note**
There is also a way to add a negative prompt with not much effort from what it looks like. Could just be added to Todos if you think it'll be worth it. I use it all the time with AUTOMATIC1111's version. I tried to implement this myself in a separate project but had no luck. I will try again though. Here is a wiki showing how it was implemented if you're interested: https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Negative-prompt

**Comment by angelmankel** [comment]
This is kind of what I had in mind.
![image](https://user-images.githubusercontent.com/23511169/193955454-8e03b08b-6c99-445e-a110-186b148ebec0.png)

**Comment by NickLucche** [comment]
> Maybe you have a preference/opinion about this but right now i'm leaning towards B.

Yep, I'd like to keep the container as simple as possible and avoid adding any unnecessary dependencies to make sure this still remains a quick way to try out stable diffusion on any machine. Your proposals are all good tho, personally I would go toward using sqlite3 which comes with Python (so it's already in the container). I can setup a few REST endpoints to read/write from it if you want. 

> But it might be worth just making a post processing tab which can include upscaling, restoration, etc. to clear up space in the editor. 

Yeah totally agree with this!

> Basically, you have a second seed which you can blend into the initial seed [...]

Okay I think we can mock that feature for now, it's not strictly "vanilla" but it also doesn't seem too hard to add.

 > There is also a way to add a negative prompt with not much effort from what it looks like

Wow that's a brilliant idea, it's really not hard to do but it's a very smart way of re-purposing the unconditional input. We can do that too, but we'll really need to reference to their work for the idea.

**Comment by angelmankel** [comment]
> personally I would go toward using sqlite3 which comes with Python (so it's already in the container). I can setup a few REST endpoints to read/write from it if you want.

Interesting, didn't know it came with Python. So yup I agree with you that sounds like a good plan to me.

> Okay I think we can mock that feature for now, it's not strictly "vanilla" but it also doesn't seem too hard to add.

Awesome!

> We can do that too, but we'll really need to reference to their work for the idea.

Couldn't agree more. I can add an about tab in the application as well.

I will go ahead and finish my mock-up of the remaining tabs and then upload them here for reference. Then i'll go ahead and start planning the React components. Shouldn't take too long to get a working prototype running. Thanks for all the input!
